### PR TITLE
Add the upgraded Zipformer model

### PR DESCRIPTION
**Pull request by yaozengwei**
This PR adds a new recipe of the upgraded Zipformer-Transducer model from @danpovey . (See https://github.com/k2-fsa/icefall/pull/1057 for detailed commit history). Compared to the old recipe ([pruned_transducer_stateless7](https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/pruned_transducer_stateless7)), the new model achieves better accuracy with lower memory usage and faster computation.

We will mainly maintain this recipe in the coming days. Other features (e.g., CTC & Attention-decoder model, multi-dataset training, language model rescoring, delay penalty, etc.) will be added into this recipe. 

Our models are trained with  **[pruned transducer loss](https://arxiv.org/abs/2206.13236)** on full librispeech dataset (with 0.9 and 1.1 speed perturbation).  We adopt **automatic mixed precision** training.
1. The results of **non-streaming** models of different scales **without LM rescoring**:
* The normal-scaled model (**65.5 M**), max-duration=1000, ~1h7m per epoch on 4 * V100-32GB gpus

| greedy_search | modified_beam_search | fast_beam_search | comment |
| - | - | - | - | 
| 2.27 & 5.1 | 2.25 & 5.06 | 2.25 & 5.04 | epoch-30-avg-9 | 
| 2.23 & 4.96 | **2.21 & 4.91** | 2.24 & 4.93 | epoch-40-avg-16 | 

* The small-scaled model (**23.2M**), max-duration=1500, ~1h30m per epoch on 2 * V100-32GB gpus
 
| greedy_search | modified_beam_search | fast_beam_search | comment |
| - | - | - | - | 
| 2.64 & 6.14 | 2.6 & 6.01 | 2.62 & 6.06 | epoch-30-avg-8 | 
| 2.49 & 5.91 | **2.46 & 5.83** | 2.46 & 5.87 | epoch-40-avg-13 | 

* The large-scaled model (**148.4M**), max-duration=1000, ~1h20m per epoch on 4 * V100-32GB gpus

| greedy_search | modified_beam_search | fast_beam_search | comment |
| - | - | - | - | 
| 2.12 & 4.91 | 2.11 & 4.9 | 2.13 & 4.93 | epoch-30-avg-9 | 
| 2.12 & 4.8 | **2.11 & 4.77**  | 2.13 & 4.78 | epoch-40-avg-16 |

2. The results of **streaming** model **without LM rescoring**: 

* The normal-scaled model (**66.1 M**), max-duration=1000, ~1h9m per epoch on 4 * V100-32GB gpus, 
* greedy_search decoding at epoch-30-avg-8
* The chunk size is at 50Hz frame rate

    - chunk size = 320ms
  
    | WER | latency | comment |
    | - | - | - | 
    | 3.21 & 8.17 | chunk-16-left-context-64 | simulated streaming (time masking) | 
    | 3.18 & 8.12 | chunk-16-left-context-64 | real streaming (chunk-wise forward) | 
    | 3.1 & 7.84 | chunk-16-left-context-128 | simulated streaming (time masking) | 
    | 3.06 & 7.78 | chunk-16-left-context-128 | real streaming (chunk-wise forward) | 

    - chunk size = 640ms
  
    | WER | latency | comment |
    | - | - | - | 
    | 2.84 & 7.24 | chunk-32-left-context-128 | simulated streaming (time masking) | 
    | 2.85 & 7.25 | chunk-32-left-context-128 | real streaming (chunk-wise forward) | 
    | 2.8 & 7.15 | chunk-32-left-context-256 | simulated streaming (time masking) | 
    | 2.84 & 7.16 | chunk-32-left-context-256 | real streaming (chunk-wise forward) |

### CODE REVIEW

**Comment by yaozengwei** [comment]
The training commands:

* normal-scaled model (65.5 M), 
```shell
./pruned_transducer_stateless7/train.py \
  --world-size 4 \
  --num-epochs 40 \
  --start-epoch 1 \
  --use-fp16 1 \
  --exp-dir pruned_transducer_stateless7/exp \
  --causal 0 \
  --full-libri 1 \
  --max-duration 1000 
```

* small-scaled model (23.2M), 
```shell
./pruned_transducer_stateless7/train.py \
  --world-size 2 \
  --num-epochs 40 \
  --start-epoch 1 \
  --use-fp16 1 \
  --exp-dir pruned_transducer_stateless7/exp-small \
  --causal 0 \
  --num-encoder-layers 2,2,2,2,2,2 \
  --feedforward-dim 512,768,768,768,768,768 \
  --encoder-dim 192,256,256,256,256,256 \
  --encoder-unmasked-dim 192,192,192,192,192,192 \
  --base-lr 0.04 \
  --full-libri 1 \
  --max-duration 1500 
```

* large-scaled model (148.4M), 
```shell
./pruned_transducer_stateless7/train.py \
  --world-size 4 \
  --num-epochs 40 \
  --start-epoch 1 \
  --use-fp16 1 \
  --exp-dir pruned_transducer_stateless7/exp-large \
  --causal 0 \
  --num-encoder-layers 2,2,4,5,4,2 \
  --feedforward-dim 512,768,1536,2048,1536,768 \
  --encoder-dim 192,256,512,768,512,256 \
  --encoder-unmasked-dim 192,192,256,320,256,192 \
  --full-libri 1 \
  --max-duration 1000 
```

* normal-scaled model, streaming (66.1 M), 
```shell
./pruned_transducer_stateless7/train.py \
  --world-size 4 \
  --num-epochs 30 \
  --start-epoch 1 \
  --use-fp16 1 \
  --exp-dir pruned_transducer_stateless7/exp-causal \
  --causal 1 \
  --full-libri 1 \
  --max-duration 1000 
```

**Comment by yaozengwei** [comment]
I use the [nvtx](https://nvtx.readthedocs.io/en/latest/index.html) profiler to visualize the event timeline for our four encoder models, including regular Conformer, reworked Conformer, old Zipformer, upgraded Zipformer, on a V100-32GB gpu. The input tensor shape of each batch is (20, 3000, 80). 

* regular Conformer, 83.5 M ([pruned_transducer_stateless](https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/pruned_transducer_stateless))
Screenshot 2023-05-13 at 19 55 13

* reworked Conformer, 77.1 M ([pruned_transducer_stateless2](https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/pruned_transducer_stateless2))
Screenshot 2023-05-13 at 19 56 07

* old Zipformer, 68.9 M ([pruned_transducer_stateless7](https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/pruned_transducer_stateless7))
Screenshot 2023-05-13 at 19 58 12

* upgraded Zipformer, 64.0 M
Screenshot 2023-05-13 at 19 57 20

File: egs/librispeech/ASR/zipformer/zipformer.py [hunk 1]
```diff
@@ -0,0 +1,2237 @@
...
+        self.conv_module1 = ConvolutionModule(embed_dim,
+                                              cnn_module_kernel,
+                                              causal=causal)
+
+        self.conv_module2 = ConvolutionModule(embed_dim,
+                                              cnn_module_kernel,
+                                              causal=causal)
+
+        # TODO: remove it
+        self.bypass_scale = nn.Parameter(torch.full((embed_dim,), 0.5))
```
Comment (marcoyang1998) [inline egs/librispeech/ASR/zipformer/zipformer.py:604]: Shall we remove this? I think it's unused.

Comment (yaozengwei) [reply egs/librispeech/ASR/zipformer/zipformer.py:604]: Yes. But the trained model has already the `bypass_scale` parameters... If we remove that, it would fail to load the model and optimizer state_dicts.

File: egs/librispeech/ASR/zipformer/zipformer.py [hunk 2]
```diff
@@ -0,0 +1,2237 @@
...
+        encoder_dim (Tuple[int]): embedding dimension of each of the encoder stacks, one per
+           encoder stack.
+        num_encoder_layers (int or Tuple[int])): number of encoder layers for each stack
+        encoder_unmasked_dim (int or Tuple[int]): unmasked dimension in each of
+            the encoder stacks for purposes of per-frame dropout (recommend 256 for
+            now).
+        query_head_dim (int or Tuple[int]): dimension of query and key per attention
+           head: per stack, if a tuple..
+        value_head_dim (int or Tuple[int]): dimension of value in each attention head
+        pos_head_dim (int or Tuple[int]): dimension of positional-encoding projection per
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/zipformer.py:69]: Please switch the order of the doc for `value_head_dim` and `pos_head_dim` to match the actual argument order.

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 3]
```diff
@@ -0,0 +1,1797 @@
...
+
+
+class PiecewiseLinear(object):
+    """
+    Piecewise linear function, from float to float, specified as nonempty list of (x,y) pairs with
+    the x values in order.  x values <[initial x] or >[final x] are map to [initial y], [final y]
+    respectively.
+    """
+    def __init__(self, *args):
+        assert len(args) >= 1
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:36]:
```suggestion
        assert len(args) >= 1, len(args)
```

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 4]
```diff
@@ -0,0 +1,1797 @@
...
+        if len(args) == 1 and isinstance(args[0], PiecewiseLinear):
+            self.pairs = list(args[0].pairs)
+        else:
+            self.pairs = [ (float(x), float(y)) for x,y in args ]
+        for (x,y) in self.pairs:
+            assert isinstance(x, float) or isinstance(x, int)
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:42]:
```suggestion
            assert isinstance(x, (float, int)), type(x)
```

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 5]
```diff
@@ -0,0 +1,1797 @@
...
+            assert isinstance(y, float) or isinstance(y, int)
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:43]:
```suggestion
            assert isinstance(y, (float, int)), type(y)
```

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 6]
```diff
@@ -0,0 +1,1797 @@
...
+
+        for i in range(len(self.pairs) - 1):
+            assert self.pairs[i + 1][0] > self.pairs[i][0], self.pairs
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:46]:
```suggestion
            assert self.pairs[i + 1][0] > self.pairs[i][0], (i, self.pairs[i], self.pairs[i + 1])
```

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 7]
```diff
@@ -0,0 +1,1797 @@
...
+                    return cur_y + (next_y - cur_y) * (x - cur_x) / (next_x - cur_x)
+                cur_x, cur_y = next_x, next_y
+            assert False
+
+    def __mul__(self, alpha):
+        return PiecewiseLinear(
+            * [(x, y * alpha) for x, y in self.pairs])
+
+    def __add__(self, x):
+        if isinstance(x, float) or isinstance(x, int):
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:71]:
```suggestion
        if isinstance(x, (float, int)):
```

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 8]
```diff
@@ -0,0 +1,1797 @@
...
+            return PiecewiseLinear(
+                * [(p[0], p[1] + x) for p in self.pairs])
+        s, x = self.get_common_basis(x)
+        return PiecewiseLinear(
+            * [(sp[0], sp[1] + xp[1]) for sp, xp in zip(s.pairs, x.pairs)])
+
+    def max(self, x):
+        if isinstance(x, float) or isinstance(x, int):
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:79]:
```suggestion
        if isinstance(x, (float, int)):
```

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 9]
```diff
@@ -0,0 +1,1797 @@
...
+                         include_crossings: bool = False):
+        """
+        Returns (self_mod, p_mod) which are equivalent piecewise lienar
+        functions to self and p, but with the same x values.
+
+          p: the other piecewise linear function
+          include_crossings: if true, include in the x values positions
+              where the functions indicate by this and p crosss.
+        """
+        assert isinstance(p, PiecewiseLinear)
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:106]:
```suggestion
        assert isinstance(p, PiecewiseLinear), type(p)
```

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 10]
```diff
@@ -0,0 +1,1797 @@
...
+
+        # get sorted x-values without repetition.
+        x_vals = sorted(set([ x for x, y in self.pairs ] + [ x for x, y in p.pairs ]))
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:109]:
```suggestion
        x_vals = sorted(set([ x for x, _ in self.pairs ] + [ x for x, _ in p.pairs ]))
```

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 11]
```diff
@@ -0,0 +1,1797 @@
...
+
+    It is a floating point value whose value changes depending on the batch count of the
+    training loop.  It is a piecewise linear function where you specifiy the (x,y) pairs
+    in sorted order on x; x corresponds to the batch index.  For batch-index values before the
+    first x or after the last x, we just use the first or last y value.
+
+    Example:
+       self.dropout = ScheduledFloat((0.0, 0.2), (4000.0, 0.0), default=0.0)
+
+    `default` is used when self.batch_count is not set or in training or mode or in
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:147]: > in training or mode

Please fix the typo.

File: egs/librispeech/ASR/zipformer/scaling.py [hunk 12]
```diff
@@ -0,0 +1,1797 @@
...
+        self.name = None
+        self.default = default
+        self.schedule = PiecewiseLinear(*args)
+
+    def extra_repr(self) -> str:
+        return f'batch_count={self.batch_count}, schedule={str(self.schedule.pairs[1:-1])}'
+
+    def __float__(self):
+        batch_count = self.batch_count
+        if batch_count is None or not self.training or torch.jit.is_scripting():
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer/scaling.py:165]: Should it be
```
not torch.jit.is_scripting()
```

Also, should we add `not torch.jit.is_tracing()`?

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/librispeech/ASR/zipformer/scaling.py (hunk 12 above)
Comment (yaozengwei) [reply egs/librispeech/ASR/zipformer/scaling.py:165]: Ok. Thanks.

**Review by csukuangfj — APPROVED** [review]

**Comment by PPGGG** [comment]
May i ask when could we use the latest zipformer ? It seems that i could not find the related code yet.

**Comment by csukuangfj** [comment]
> May i ask when could we use the latest zipformer ? 

Whenever you want.

Please see https://github.com/k2-fsa/sherpa/pull/379

@PPGGG

**Comment by PPGGG** [comment]
> > May i ask when could we use the latest zipformer ?
> 
> Whenever you want.
> 
> Please see [k2-fsa/sherpa#379](https://github.com/k2-fsa/sherpa/pull/379)
> 
> @PPGGG

Yeah that's great. But i would like to find the training recipe to train the lastest zipformer on my own dataset. Should i just use the `zipformer` folder under `librispeech/ASR` ? Thanks.

**Comment by csukuangfj** [comment]
Yes, you can find the usage in RESULTS.md

**Comment by JinZr** [comment]
hi, you can find the code for simulated streaming decode in the librispeech zipformer recipe, simulated streaming is enabled by setting causal to 1. also remember to set a proper value for `left-context-frames` and `chunk-size`.

best
jin

**Comment by Dokotela** [comment]
Hello, I realize this is not a new model, but was still hoping to ask a question. I've been working with the sherpa-onnx version of the model in this PR, and comparing it to the sherpa-onnx version of [this model](https://github.com/k2-fsa/icefall/pull/984). I work in healthcare, so I'm looking to try and tune a model for medical dictation (especially for emergency and disaster relief scenarios). The older version works notably better than this one when I've tested it on my datasets (which have medical lectures, podcasts, some actual doctor-patient interactions, and some tv dialogue from medical shows). The other model runs a WER of around 12%, but this model runs over 20%.  I was wondering, is this possibly from something in the architecture itself, or more likely from the dataset it was trained on? (this model looks like it only used librispeech, but the other model used librispeech and gigaspeech). I was curious if anyone had any ideas, or if someone else had run across this difference previously.

**Comment by JinZr** [comment]
i think it's mostly due to the difference in the training data, the
librispeech dataset contains only literature reading while the gigaspeech
one contains speech collected in-the-wild, with a broader spectrum of
topics. plus the gigaspeech dataset is significantly larger than
librispeech.
